File: java/gov/nasa/gsfc/gmsec/api/util/TimeUtil.java
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file TimeUtil.java
 */

package gov.nasa.gsfc.gmsec.api.util;

import gov.nasa.gsfc.gmsec.api.jni.util.JNITimeUtil;


/**
 * A Time Utility class
 */
public class TimeUtil
{
	// To prevent users from instantiating this class.
	private TimeUtil() {}


	/**
	 * Returns the current time in seconds since Jan 1, 1970 00:00::00 UTC.
	 *
	 * @return A TimeSpec object.
	 */
	public static TimeSpec getCurrentTime()
	{
		return JNITimeUtil.getCurrentTime();
	}


	/**
	 * Returns a GMSEC time string YYYY-DDD-HH-MM-SS.sss for the given TimeSpec.
	 *
	 * @param spec A TimeSpec time
	 *
	 * @return A time string.
	 *
	 * @see TimeUtil#getCurrentTime()
	 */
	public static String formatTime(TimeSpec spec)
	{
		if (spec == null)
		{
			throw new IllegalArgumentException("TimeSpec is null");
		}

		return JNITimeUtil.formatTime(spec);
	}


	/**
	 * Returns a GMSEC time string YYYY-DDD-HH-MM-SS[....sss] for the given TimeSpec.
	 *
	 * @param spec       A TimeSpec time
	 * @param subseconds The number of digits to allow for subseconds [0 .. 9].
	 *
	 * @return A time string.
	 *
	 * @see TimeUtil#getCurrentTime()
	 */
	public static String formatTime(TimeSpec spec, int subseconds)
	{
		if (spec == null)
		{
			throw new IllegalArgumentException("TimeSpec is null");
		}
		if (subseconds < 0 || subseconds > 9)
		{
			throw new IllegalArgumentException("Subsecond value is out of range");
		}

		return JNITimeUtil.formatTime(spec, subseconds);
	}
}

File: framework/src/gmsec4/c/util/timeutil.cpp
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */



/** @file timeutil.cpp
 *
 *  @brief This file contains C-binding time utility functions.
 */

#include <gmsec4/c/util/timeutil.h>

#include <gmsec4/util/TimeUtil.h>


void CALL_TYPE timeUtilMillisleep(unsigned int milliseconds)
{
	gmsec::api::util::TimeUtil::millisleep(milliseconds);
}


GMSEC_TimeSpec CALL_TYPE timeUtilGetCurrentTime()
{
	return gmsec::api::util::TimeUtil::getCurrentTime();
}


double CALL_TYPE timeUtilGetCurrentTime_s(GMSEC_TimeSpec* ts)
{
	return gmsec::api::util::TimeUtil::getCurrentTime_s(ts);
}


GMSEC_TimeSpec CALL_TYPE timeUtilGetCharTime(const char* timeString)
{
	return gmsec::api::util::TimeUtil::getCharTime(timeString);
}


double CALL_TYPE timeUtilGetCharTime_s(const char* timeString)
{
	return gmsec::api::util::TimeUtil::getCharTime_s(timeString);
}


void CALL_TYPE timeUtilFormatTime(const GMSEC_TimeSpec time, GMSEC_Time buffer)
{
	gmsec::api::util::TimeUtil::formatTime(time, buffer);
}


void CALL_TYPE timeUtilFormatTimeSubs(const GMSEC_TimeSpec time, int subs, GMSEC_Time buffer)
{
	gmsec::api::util::TimeUtil::formatTime(time, subs, buffer);
}


GMSEC_TimeSpec CALL_TYPE timeUtilDiffTime(const GMSEC_TimeSpec time1, const GMSEC_TimeSpec time0)
{
	return gmsec::api::util::TimeUtil::diffTime(time1, time0);
}


GMSEC_BOOL CALL_TYPE timeUtilEqualTo(const GMSEC_TimeSpec time1, const GMSEC_TimeSpec time0)
{
	return (time1 == time0 ? GMSEC_TRUE : GMSEC_FALSE);
}


GMSEC_BOOL CALL_TYPE timeUtilGreaterThan(const GMSEC_TimeSpec time1, const GMSEC_TimeSpec time0)
{
	return (time1 > time0 ? GMSEC_TRUE : GMSEC_FALSE);
}


GMSEC_BOOL CALL_TYPE timeUtilLesserThan(const GMSEC_TimeSpec time1, const GMSEC_TimeSpec time0)
{
	return (time1 < time0 ? GMSEC_TRUE : GMSEC_FALSE);
}
